### ISSUE TITLE
Should the project switch to a different code / style / enforcer

### ISSUE THREAD

**Issue by mfn** [rebing/graphql-laravel#497]
Spin-off from https://github.com/rebing/graphql-laravel/pull/485#issuecomment-544126161

When I started to increase my contributions to the project, my first and biggest gripe was that there was not coherent code style in the code base.

Having had experience with https://styleci.io/ , I simple went ahead to "fix it quickly" and ensured that all source files where on the same level. An additional benefit of this integration was that, independent on who contributes, the style is implicitly enforced as a dedicated check which needs to be green before a merge can happen; ensuring the code base style stays coherent..

From the referenced issue:
> Do you have some command to verify the style code in my local environment? Something like that phpcs --standard=phpcs.xml --extensions=php src.

My answer:
> Unfortunately no, StyleCI entirely works "in the cloud"; so only after a push you get back the information via Github. The "best" you can do then, instead of manually fixing it, is to simply download the diff and apply it, push again (or, as I usually do it, rebase)

### TL;DR: I acknowledge this is not ideal and I'm open for changing this

Things to be considered for an alternative:
- [x] preferable open source, as is this package 😄 
- [x] easy to run locally for everyone
- [ ] be able to automatically commit the fixes via a CI pipeline (github actions?)

The astute reader will notice the above points don't talk about _the actual code style_. For a start, no one specifically complained about it. And personally I don't care _too much_ about which actual style, as long as some criteria can be met:
- should be a well acknowledged / used style in the OSS community
  Currently it's using the "Laravel" style guide as implemented by StyleCI; it being simply a Laravel package. Personally though I've no feelings that it has to be "the Laravel" style
- should be enforceable within the IDE automatically "to a high degree"
  I'm fine that an external / OSS tool can be used to always align the source to the "final form", but an IDE should be able to "come close" to it so ideally the necessary changes aren't too much every time
  My experience with the current setup and e.g. PhpStorm is quite a good one. In 99% of the cases of StyleCI it's usually only some `use` statement reordering when I work with it

Feedback / actionables very much welcome. Personally outside StyleCI I've not much experience so basically this issue relies on someone fill the void a bit here :)

**Comment by crissi** [comment]
using phpfixer with @PSR2 rule set would be nice supplemented with what ever custom rule makes sense.

But most importantly the possiblity to run it locally.

Another alternative is for the pipeline to fix the styling and commit on push.

**Comment by mfn** [comment]
> using phpfixer with @PSR2 rule set would be nice supplemented with what ever custom rule makes sense.

If possible, I would like to do as few customization as necessary. Less discussion, less surprise.

> But most importantly the possiblity to run it locally.
> 
> Another alternative is for the pipeline to fix the styling and commit on push.

Can't we have both? Something people can run locally but also have a push pipeline, in case it doesn't match?

**Comment by crissi** [comment]
Yes using php fixer this should not be a problem

**Comment by mfn** [comment]
Awesome!

How can we do this? :

**Comment by crissi** [comment]
Never played with github pipelines, but I can try to cook something up

Hent Outlook til iOS<https://aka.ms/o0ukef>
________________________________
Fra: Markus Podar <notifications@github.com>
Sendt: Wednesday, October 23, 2019 1:22:11 AM
Til: rebing/graphql-laravel <graphql-laravel@noreply.github.com>
Cc: Christian <c2n@live.dk>; Comment <comment@noreply.github.com>
Emne: Re: [rebing/graphql-laravel] Should the project switch to a different code / style / enforcer (#497)


Awesome!

How can we do this? :

**Comment by mfn** [comment]
Thanks to @crissi we've [the new style system ](https://github.com/rebing/graphql-laravel/pull/502)in-place 🤗 

I'll still leave this issue open for now as we've no automatic-commit solution yet, which I still would like to see as it saves the tedious work of developers having to think about running the fixer.

**Comment by georgeboot** [comment]
> I'll still leave this issue open for now as we've no automatic-commit solution yet
Do you mean you want something similar to StyleCI that auto applies the fixes?

A cloud solution might be to create a GitHub Action that runs the fixer and commits it back to the repo.

If you want it on the developers laptop, you could perhaps create a git hook that will run it?

**Comment by mfn** [comment]
> Do you mean you want something similar to StyleCI that auto applies the fixes?

Yes!

I'm also thinking "going Github Actions" and what I read, it should be _rather_ easy.

Currently this is on hold because AFAIK we still have the StyleIC integration which basically totally clashes with the PHP-CS fixer :/

**Comment by georgeboot** [comment]
Committing from an action on its own is easy. 

However, there are some potential issues:

1. You need a user to commit with. Since you don't have the access details of the real user, you need to use @rebing 's account.
2. If the creator of the PR doesn't allow access to repo admins, you can't commit in his repo.

Perhaps it would be easier to write a simple Bot, since they can commit on behalf of users. But now you need to host this bot yourself etc... Not really worth it.

What this all boils down to, what was exactly the issue with StyleCI? If StyleCI's bot automatically fixes the style issues (it can do that), there is no real downside other than having to pull the fixed code before you can push new commits to the PR branch, right?

**Comment by mfn** [comment]
> what was exactly the issue with StyleCI?

People couldn't run it locally upfront and it was a hassle to "fix" the StyleCI errors because few realized you could download a diff and apply to the source. But that is also an annoying workflow.

TL;DR: we didn't enable auto-commit for StyleCI but it would be great if we could provide this features, saves everyone time 😄 

> you need to use @rebing 's account.

Are you sure? My understand of github permission was that it could also be a token from my user because, in that regard, I've also permissions to push if the remote allows.

_However_ you're right, it's sure best it's from a repo admin (which I'm not).

OTOH, I'm not concerned if pushes won't work in certain cases. a) we tried 😄  and b) people can run it on their own anyway

**Comment by georgeboot** [comment]
> TL;DR: we didn't enable auto-commit for StyleCI but it would be great if we could provide this features, saves everyone time 😄

So you are basically saying that we should just enable auto-patch and then everyone will be happy? No more manual patching fixes, no more manual running. Just push and go.

> > you need to use @rebing 's account.
> 
> Are you sure? My understand of github permission was that it could also be a token from my user because, in that regard, I've also permissions to push if the remote allows.

Yes, Actions is very strict in this regard. It won't allow you to commit on someones behalf. This would potentially allow contributors to PR a new actions script that would write to master etc.

**Comment by mfn** [comment]
> Yes, Actions is very strict in this regard. It won't allow you to commit on someones behalf. This would potentially allow contributors to PR a new actions script that would write to master etc.

I guess you're right. I was thinking about github token being provided via the secrets, but then again this is only something admin can do anyway, so 🤷‍♀️

> So you are basically saying that we should just enable auto-patch and then everyone will be happy? No more manual patching fixes, no more manual running. Just push and go.

Well, if it works, it saves time 😄 At least that was my simple thinking.

If others disagree, of course it's must and we can close this issue.

**Comment by mfn** [comment]
Closing this issue:
- style checking has been moved to php-cs fixer
- styleci integration has been removed
- auto-push of fixes isn't a hard requirement (though IMHO it would still be nice)

Thanks especially @crissi for the work on it 🙇
